sourcegraph/sourcegraph
# [Cody Gateway] add refresh token endpoint with sams authenticator

This is the first of many upcoming PRs to enable refreshing user's rate limits on Cody Gateway when a subscription change happens on SSC.

Issue: https://github.com/sourcegraph/accounts.sourcegraph.com/issues/363
Related SAMS PR: https://github.com/sourcegraph/accounts.sourcegraph.com/pull/418

**The whole plan is as follows (given we do not have SAMSActor on Gateway yet):**
- On every subscription change, SSC will make a request to Cody Gateway to refresh the rate limits for the user, the same as dotcom does right now.
- SSC will send dotcom_user_id with the request, which it will get from the user's metadata from SAMS. 
- Dotcom will expose a new graphql resolver which provides rate limits for user ID. Right now, it provides one, but for a dotcom access token.
- Gateway will make a call to dotcom to fetch the limits for the dotcom_user_id using the new API. It does the same now but for dotcom's per_user_access_token.

**How will it look when we have SAMSActor & SSCSource on Gateway?**
Same as above, but with a few changes:
- SSC will send SAMS user_access_token in place of dotcom_user_id with the request.
- SSC will expose a graphql resolver which provides rate limits for SAMS user_access_token.
- Gateway will use SAMS user_access_token to fetch rate limits directly from SSC.

**What does this PR handle?**
1. It implements a new SAMS client.
2. It implements a new SAMS auth middleware, which allows auth using service-to-service tokens utilizing the SAMS client.
3. It exposes a new dummy API endpoint, `/v1/sams/limits/refresh`, which utilises the new SAMS auth middleware. To note, the `/v1/limits/refresh` endpoint already exists, but it uses the present auth. 

**What does this PR NOT handle? (TODOs for upcoming PRs)**
1. The actual implementation of the new API endpoint is pending and will be addressed in upcoming PRs. 
2. It doesn't update the dotcom graphql API either, which will be addressed in upcoming PRs.

## Test plan
- Checkout SAMS local on https://github.com/sourcegraph/accounts.sourcegraph.com/pull/418
- Run SAMS locally under the 9982 port. Otherwise, it will conflict with Gateway running on 9992.
- Create client ID and client secrets using: 
![CleanShot 2024-02-01 at 21 53 45@2x](https://github.com/sourcegraph/sourcegraph/assets/22571395/02afcdd9-5da4-4590-ac4e-e7a37a68f3d4)
- Update scopes to include `cody.gateway` in the DB directly `psql -U sourcegraph accounts` `update idp_clients set scopes = '["openid", "profile", "email", "offline_access", "cody.gateway", "cody.ssc"]';`
- Update `sg.config.overwrite.yaml` in sg repo to include the client secret and id similar to this: 
![CleanShot 2024-02-01 at 21 56 08@2x](https://github.com/sourcegraph/sourcegraph/assets/22571395/88ad7867-7f4e-46eb-b186-c091a79a27f1)
- Update `.envrc` in accounts repo to include include the client secret, id and gateway URL similar to this: 
![CleanShot 2024-02-01 at 21 57 19@2x](https://github.com/sourcegraph/sourcegraph/assets/22571395/28476635-3bbc-45be-864b-9b3e5c2d8416)
- Search for codyAdminAccountIDs in accounts repo and copy the admin account id hard coded. Update  your the `users.external_id` and `users.accounts_external_id` in accounts and cody_management DB respectively, which the admin account id. 
- Visit `localhost:9982` in the browser. Make sure you are logged in with the account you set as admin above. 
- Make the following request from the console to test the flow: 
```
await (await fetch(`${window.location.href}cody/api/rest/user/{put the external_id here}/limits/refresh`, { method: 'POST' })).json()
```
